src/neo/Ledger/Blockchain.cs

```diff
@@ -263,22 +266,57 @@ private void OnImport(IEnumerable<Block> blocks, bool verify)
             Sender.Tell(new ImportCompleted());
         }
 
+        private void RemoveUnverifiedBlockToCache(uint index, bool removeKnownHashes)
+        {
+            if (block_cache_unverified.Remove(index, out var entry))
+            {
+                block_cache_unverified_size -= entry.Size;
+
+                if (removeKnownHashes)
+                {
+                    var hashes = entry.Blocks.Select(u => u.Hash).ToArray();
+
+                    system.LocalNode.Tell(new TaskManager.RemoveKnownHashes { Hashes = hashes });
+                    system.TaskManager.Tell(new TaskManager.RemoveKnownHashes { Hashes = hashes });
+                }
+            }
+        }
+
         private void AddUnverifiedBlockToCache(Block block)
         {
+            if (block_cache_unverified_size > MaxUnverifiedBlockSize && block.Index > Height + 10)
+            {
+                // Drop last entry if it's higher
+
+                if (block_cache_unverified.Count > 0)
+                {
+                    var max = block_cache_unverified.Keys.Max();
+                    if (max > Height + 10)
+                    {
+                        RemoveUnverifiedBlockToCache(max, true);
+                    }
+                }
+
+                // Drop block, we can't save it safely in memory
+                return;
```

Comment:
Drop both earlier and new coming blocks?

Reply:
Fixed